Documentation/bpf.rst

```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
+  end remains in the initial namespace of the host, all network traffic from the
+  container has to pass through the host-facing veth device allowing for attaching
+  tc BPF programs on the tc ingress and egress hook of the veth. Network traffic
+  going into the container will pass through the host-facing veth's tc egress
+  hook whereas network traffic coming from the container will pass through the
+  host-facing veth's tc ingress hook.
+
+  For virtual devices like veth devices XDP is unsuitable in this case since the
+  kernel operates solely on a ``skb`` here and generic XDP has a few limitations
+  where it does not operate with cloned ``skb``'s. The latter is heavily used
+  from the TCP/IP stack in order to hold data segments for retransmission where
+  the generic XDP hook would simply get bypassed instead. Moreover, generic XDP
+  needs to linearize the entire ``skb`` resulting in heavily degraded performance.
+  tc BPF on the other hand is more flexible as it specializes on the ``skb``
+  input context case and thus does not need to cope with the limitations from
+  generic XDP.
+
+..
+
+* **Forwarding and load-balancing**
+
+  The forwarding and load-balancing use case is quite similar to XDP, although
+  slightly more targeted towards east-west container workloads rather than
+  north-south traffic (though both technologies can be used in either case).
+  Since XDP is only available on ingress side, tc BPF programs allow for
+  further use cases that apply in particular on egress, for example, container
+  based traffic can already be NATed and load-balanced on the egress side
+  through BPF out of the initial namespace such that this is done transparent
+  to the container itself. Egress traffic is due to the nature of the kernel's
+  networking stack already ``skb`` based anyway such that packet rewrites and
+  redirects are suitable out of tc BPF. Through utilizing the ``bpf_redirect()``
+  helper function, BPF can take over the forwarding logic to push the ``skb``
+  either into the ingress or egress path of another networking device. Thus,
+  any bridge-like devices become unnecessary to use as well by utilizing
+  tc BPF as forwarding fabric.
+
+..
+
+* **Flow sampling, monitoring**
+
+  Like in XDP case, flow sampling and monitoring can be realized through a
+  high-performance lockless per-CPU memory mapped perf ring buffer where the
+  BPF program is able to push custom data, the full or truncated packet
+  contents, or both up to a user space application. From the tc BPF program
+  this is realized through the ``bpf_skb_event_output()`` BPF helper function
+  which has the same function signature and semantics as ``bpf_xdp_event_output()``.
+  Given tc BPF programs can be attached to ingress and egress as opposed to
+  only ingress in XDP BPF case plus the two tc hooks are at the lowest layer
+  in the (generic) networking stack, this allows for bidirectional monitoring
+  of all network traffic from a particular node.
+
```

Comment:
Could be worth mentioning here that this harkens back to old-school BPF and tcpdump.